README.md
# Clojure Lab

<img src="https://raw.github.com/jfacorro/clojure-lab/master/resources/logo.png" align="right" style="float:right" />

**Clojure Lab** is a project that aims to build a development environment for Clojure in Clojure, its main goals are *usability* and *extensibility*. This project is currently under development, so most of the assertions that follow are closer to wishful thinking than reality... for now :).

### Usability

- When new to the environment you should feel comfortable while discovering and interacting with the features available.
- There should be no road blocks and work should flow naturally once you are familiarized with the way things work.
- Enable the best possible use of the keyboard and mouse combination, both are useful tools for interaction and they each should have their place while working in the environment.

### Extensibility

- Provide a simple API to create your own tools and controls that can be used in the environment, for example in case there's a tool that you are missing and you feel like scratching your own itch.
- New types of projects and documents can be created with plugins.
- Plugins will have access to every core library, so that existing functionality can be enhanced and built upon.

## Rationale

When I started using Clojure I bumped into the situation where I didn't quite know what to use for Clojure development. [Emacs][6] seemed too big a challenge to start playing with a new language, so while reading the [Joy of Clojure][7] and going through its examples, [Clooj][4] was a pretty good tool to explore the basics. Although Clooj has its shortcomings it provides a friendly, simple and familiar interface.

After some time of Clojure development three things happened that brought **Clojure Lab** into being:

- Having to present a final project for me to graduate from school.
- Reading [Chas Emerick][2]'s post [The Ideal Clojure Development Environment][1].
- Thinking that I didn't have an accessible choice for a Clojure development environment that would scale for bigger projects.

Even when I discovered [Counterclockwise][3] and [Light Table][5] was annouced , I still decided to try and build an IDE that could cover as many of the features mentioned in [The Ideal Clojure Development Environment][1] as possible. The final goal is to have a powerful extensible tool, yet simple and accessible, that provides a reactive user interface with discoverable features (a lot easier said than done, right?).

## Running the Application

### Requirements

In order to fire up Clojure Lab from the REPL, build or run it you need:

- [Leiningen](https://github.com/technomancy/leiningen/) installed and in your `$PATH`.
- [JDK 1.7.0](http://www.oracle.com/technetwork/java/javase/downloads/jdk7-downloads-1880260.html) installed.

### REPL

The application can be run by either using the `lein run` or starting a REPL and then calling the `-main` function:

    $ lein repl
    lab.main> (-main)

### Building

An executable `lab.jar` file can be generated through the use of the `lein build` command. 

### Downloading

Pre-built executable jar files are available in the [releases][9] section of this project.

## User Guide

There's a small user manual/guide that shows how to start using **Clojure Lab** [here][8].

## Features

Most of the features in the following list were extracted from the post [The Ideal Clojure Development Environment][1]:

### Code Editing

- Syntax highlighting. **DONE**
- Brace, bracket, paren matching. **DONE**
- Paredit equivalency. **DONE**
- Easily-togglable rainbow parens. **DONE**
- S-expression navigation.
- In-place popup macroexpansion. **DONE**
- Auto completion. **DONE**

### Project Organization

- File management. **DONE**
- Code compilation.
- Dependencies resolution (maven & leiningen support).
- Initial configuration. **DONE**
- Static analysis:
    - Current file code outline. **DONE**
    - Static namespace browser.

### REPL
- Multiple REPLs support (each running on a separate process). **DONE**
- Execution history:
    - Search.
    - Execute again. **DONE**
- Runtime namespace browser.
- Full editor capability in the REPL.
- Automatic generation and configuration of the classpath for local REPLs. **DONE (Leiningen)**

### Nice-to-have
- Full Java support Integrated (debugging, code completion, profiling, etc.).
- Code generation (deftype/defrecord/extend-type/gen-class/proxy).
- REPL
    - Configurable pretty-printing of output.
    - Being able to `print` non-textual data, like images and such.
- Static analysis
    - Symbol navigation (i.e. "Go to declaration").
    - Find usages.

# License

Copyright © 2014 Juan Facorro

Distributed under the Eclipse Public License, the same as Clojure.

  [1]: http://cemerick.com/ideal-clojure-development-environment/
  [2]: http://cemerick.com/
  [3]: https://code.google.com/p/counterclockwise/
  [4]: https://github.com/arthuredelstein/clooj
  [5]: http://www.lighttable.com/
  [6]: http://www.gnu.org/software/emacs/
  [7]: http://joyofclojure.com/
  [8]: https://github.com/jfacorro/clojure-lab/blob/master/docs/manual.md
  [9]: https://github.com/jfacorro/clojure-lab/releases


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/jfacorro/clojure-lab/tree/HEAD)

[Full Changelog](https://github.com/jfacorro/clojure-lab/compare/v0.1.0-beta...HEAD)

**Fixed bugs:**

- Update User Manual images urls and complete REPL documentation [\#65](https://github.com/jfacorro/clojure-lab/issues/65)

**Closed issues:**

- Markdown: add shortcuts for links and keystrokes [\#62](https://github.com/jfacorro/clojure-lab/issues/62)

- User Manual [\#61](https://github.com/jfacorro/clojure-lab/issues/61)

**Merged pull requests:**

- \[Closes \#65\] Change URLs [\#66](https://github.com/jfacorro/clojure-lab/pull/66) ([jfacorro](https://github.com/jfacorro))

- \[Closes \#61\] User Manual. [\#64](https://github.com/jfacorro/clojure-lab/pull/64) ([jfacorro](https://github.com/jfacorro))

- \[Closes \#62\] Added shortcuts. Fixed deps issue. [\#63](https://github.com/jfacorro/clojure-lab/pull/63) ([jfacorro](https://github.com/jfacorro))

## [v0.1.0-beta](https://github.com/jfacorro/clojure-lab/tree/v0.1.0-beta) (2014-06-27)

[Full Changelog](https://github.com/jfacorro/clojure-lab/compare/v0.1.0-alpha...v0.1.0-beta)

**Implemented enhancements:**

- Create a short manual/guide [\#51](https://github.com/jfacorro/clojure-lab/issues/51)

- After saving a doc, if the language changes update the view [\#47](https://github.com/jfacorro/clojure-lab/issues/47)

- Show "Find in Files" results tree expanded by default [\#44](https://github.com/jfacorro/clojure-lab/issues/44)

- Improve information displayed by contextual help [\#43](https://github.com/jfacorro/clojure-lab/issues/43)

- Inline docstrings [\#42](https://github.com/jfacorro/clojure-lab/issues/42)

- Implement a way to specify the order of the menu items when defining a :global keymap [\#39](https://github.com/jfacorro/clojure-lab/issues/39)

- Remove confirmation before exiting... [\#37](https://github.com/jfacorro/clojure-lab/issues/37)

- Modify the key used in the :lang-plugins configuration map. [\#27](https://github.com/jfacorro/clojure-lab/issues/27)

- Some plugins should be activated explicitly... [\#23](https://github.com/jfacorro/clojure-lab/issues/23)

**Fixed bugs:**

- Markdown preview plugin is not being loaded [\#48](https://github.com/jfacorro/clojure-lab/issues/48)

- ui: there is a white line in the top of all :tabs components [\#46](https://github.com/jfacorro/clojure-lab/issues/46)

- CompilerException on Ubuntu 14.04 [\#45](https://github.com/jfacorro/clojure-lab/issues/45)

- UI - Make local attr override any attr set through a stylesheet [\#41](https://github.com/jfacorro/clojure-lab/issues/41)

- lab.core.keymap/commands should return only top-most bindings [\#38](https://github.com/jfacorro/clojure-lab/issues/38)

- Find & Replace: searching for the empty string causes StackOverflow. [\#33](https://github.com/jfacorro/clojure-lab/issues/33)

- bundle-operations macro fails with "Too many arguments to if" [\#32](https://github.com/jfacorro/clojure-lab/issues/32)

- Search & Open File doesn't filter out directories [\#30](https://github.com/jfacorro/clojure-lab/issues/30)

- paredit: closing delimiter not inserted [\#29](https://github.com/jfacorro/clojure-lab/issues/29)

- nREPL plugin: lein command not being found in linux [\#28](https://github.com/jfacorro/clojure-lab/issues/28)

- Global plugin commands are not being loaded as menu items [\#26](https://github.com/jfacorro/clojure-lab/issues/26)

- After opening search dialog, close button behavior from tabs changes [\#25](https://github.com/jfacorro/clojure-lab/issues/25)

- Some plugins should be activated explicitly... [\#23](https://github.com/jfacorro/clojure-lab/issues/23)

**Closed issues:**

- paredit: unbalanced delete with strings [\#31](https://github.com/jfacorro/clojure-lab/issues/31)

- dependencies badge [\#11](https://github.com/jfacorro/clojure-lab/issues/11)

## [v0.1.0-alpha](https://github.com/jfacorro/clojure-lab/tree/v0.1.0-alpha) (2014-03-07)

[Full Changelog](https://github.com/jfacorro/clojure-lab/compare/prototype...v0.1.0-alpha)

**Implemented enhancements:**

- Improve performance of rainbow delimiters plugin [\#24](https://github.com/jfacorro/clojure-lab/issues/24)

- Implement lazy loading for the file explorer [\#14](https://github.com/jfacorro/clojure-lab/issues/14)

- Implement stylesheets [\#7](https://github.com/jfacorro/clojure-lab/issues/7)

- lab.ui.select - Implement a way to get ALL components matching a selector [\#5](https://github.com/jfacorro/clojure-lab/issues/5)

- Close button from doc tab shows image with border from button \(Ubuntu\) [\#3](https://github.com/jfacorro/clojure-lab/issues/3)

**Fixed bugs:**

- Cancelling the save dialog closes a modified not saved document [\#22](https://github.com/jfacorro/clojure-lab/issues/22)

- paredit: open-delimiter in comment inserts a closing delimiter... [\#21](https://github.com/jfacorro/clojure-lab/issues/21)

- paredit: wrap-around generates NPE when caret position is at the end of the text [\#20](https://github.com/jfacorro/clojure-lab/issues/20)

- Application freezes when on Clojure document and entering \#" [\#19](https://github.com/jfacorro/clojure-lab/issues/19)

- On click of tree items throws NPE when none is selected [\#18](https://github.com/jfacorro/clojure-lab/issues/18)

- NullReferenceException when using the menu's Save command... [\#17](https://github.com/jfacorro/clojure-lab/issues/17)

- NullReferenceException when cancelling file explorer dialog [\#16](https://github.com/jfacorro/clojure-lab/issues/16)

- Opening a file from the file explorer doesn't work... [\#15](https://github.com/jfacorro/clojure-lab/issues/15)

- dbl-click on the file explorer with no selection results in NullPointerException [\#13](https://github.com/jfacorro/clojure-lab/issues/13)

- Error when running the Gtk test \(lab.ui.gtk.test\) [\#10](https://github.com/jfacorro/clojure-lab/issues/10)


src/lab/ui.clj
(ns lab.ui
  "DSL to abstract the UIcomponents with Clojure data structures."
  (:require [lab.ui [core :as ui]
                    [select :as ui.sel]
                    [tree :as tree]
                    [menu :as menu]
                    [stylesheet :as css]
                    [protocols :as p]
                    swing]
             [lab.core [keymap :as km]
                       [plugin :as plugin]]
             [lab.model.document :as doc]))

(declare document-tab)

(defn- current-document-tab [ui]
  (-> @ui (ui/find :#documents) p/selected))

; Open

(defn- open-document-ui
  "Used by the open and new commands."
  [app doc]
  (as-> (:ui app) ui
    (ui/update! ui :#documents p/add (document-tab app doc))))

(defn- open-document-menu
  [app evt]
  (let [file-dialog   (ui/init [:file-dialog {:type :open}])
        [result file] (p/show file-dialog)]
    (if file
      (lab.app/open-document app (.getCanonicalPath file))
      app)))

(defn- open-document-hook
  [f app path]
  (let [app (f app path)
        doc (lab.app/current-document app)]
    (open-document-ui app doc)
    app))

; New

(defn- new-document-hook
  [f app & evt]
  (let [app  (f app)
        doc  (lab.app/current-document app)]
    (open-document-ui app doc)
    app))

; Close

(defn close-tab
  [ui id & _]
  (let [tab  (ui/find @ui (str "#" id))]
    (ui/update! ui :#documents p/remove tab)))

(defn- close-document-hook
  "Finds the currently selected tab, removes it and closes the document
associated to it."
  [f app & evt]
  (let [ui    (:ui app)
        tab   (current-document-tab ui)
        doc   (ui/get-attr tab :doc)
        app   (f app @doc)]
    (ui/update! ui :#documents p/remove tab)
    app))

; Save

(defn- assign-path
  "If the document doesn't have a path, get one from the user."
  [doc]
  (if (doc/path doc)
    doc
    (let [file-dialog   (ui/init [:file-dialog {:type :save}])
          [result file] (p/show file-dialog)]
      (if file
        (doc/bind doc (.getCanonicalPath file) :new? true)
        doc))))

(defn- save-document-hook
  [f app & evt]
  (let [ui    (:ui app)
        tab   (current-document-tab ui)
        doc   (ui/get-attr tab :doc)]
    (swap! doc assign-path)
    (when (doc/path @doc)
      (f app doc))))

; Insert

(defn text-editor-change [app doc evt]
  (case (:type evt)
    :insert (swap! doc doc/insert (:offset evt) (:text evt))
    :remove (swap! doc doc/delete (:offset evt) (+ (:offset evt) (:length evt)))
    :change nil))

; Register

(defn- register-keymap-hook
  [f app keymap]
  (case (:type keymap)
    :global
      (let [ui    (:ui @app)
            cmds  (-> keymap :bindings vals)]
        (swap! ui (partial reduce (partial menu/add-option app)) cmds))
     :lang  nil
     :local nil)
  (f app keymap))

(defn- create-text-editor [app doc]
  [:text-editor {:text        (doc/text @doc)
                 :border      :none
                 :background  0x666666
                 :foreground  0xFFFFFF
                 :caret-color 0xFFFFFF
                 :on-change   (partial #'text-editor-change app doc)
                 :font        [:name "Monospaced.plain" :size 14]}])

(defn- document-tab [app doc]
  (ui/with-id id
    [:tab {:doc      doc
           :tool-tip (doc/path @doc)
           :header   [:panel {:transparent true}
                             [:label {:text (doc/name @doc)}]
                             [:button {:icon         "close-tab.png"
                                       :border       :none
                                       :transparent  true
                                       :on-click     (partial #'close-tab (:ui app) id)}]]
           :border    :none}
           (create-text-editor app doc)]))

(defn build-main [app-name]
  [:window {:id     "main"
            :title   app-name
            :size    [700 500]
            :maximized true
            :icons   ["icon-16.png" "icon-32.png" "icon-64.png"]
            :menu    [:menu-bar]}
            [:split {:orientation :vertical
                     :resize-weight 1
                     :border :none}
                    [:split {:resize-weight 0}
                            [:tabs {:id "left-controls"}]
                            [:split {:resize-weight 1}
                                     [:tabs {:id "documents"}]
                                     [:tabs {:id "right-controls"}]]]
                    [:tabs {:id "bottom-controls"}]]])

(defn- open-document-tree
  "Handler for the click event of an item in the tree."
  [app {:keys [source click-count]}]
  (when (= click-count 2)
    (let [^java.io.File file (p/selected source)]
      (when-not (.isDirectory file)
        (lab.app/open-document app (.getCanonicalPath file))))))

(defn- file-tree [app]
  [:tab {:title "Files" :border :none}
        [:tree {:id      "file-tree"
                :on-click (partial #'open-document-tree app)
                :root     (tree/load-dir "/home/jfacorro/dev/clojure-lab/src/lab/ui/swing")}]])

(defn add-component
  "Add component to a tabs control."
  [app selector title component]
  (let [tab      [:tab {:title title :border :none} component]]
    (ui/update! (app :ui) selector p/add component)))

(def hooks
  {#'lab.core.plugin/register-keymap! #'register-keymap-hook
   #'lab.app/new-document             #'new-document-hook
   #'lab.app/open-document            #'open-document-hook
   #'lab.app/save-document            #'save-document-hook
   #'lab.app/close-document           #'close-document-hook})

(def keymaps
  [(km/keymap (ns-name *ns*)
              :global
              {:category "File" :name "New" :fn #'lab.app/new-document :keystroke "ctrl N"}
              {:category "File" :name "Open" :fn #'open-document-menu :keystroke "ctrl O"}
              {:category "File" :name "Close" :fn #'lab.app/close-document :keystroke "ctrl W"}
              {:category "File" :name "Save" :fn #'lab.app/save-document :keystroke "ctrl S"})])

(defn init!
  "Expects an atom containing the app. Builds the basic UI and 
  adds it to the app under the key :ui."
  [app]
  (let [ui (atom (-> (:name @app) build-main ui/init))]
    (swap! app assoc :ui ui)
    
    (add-component @app :#left-controls "Files" (file-tree @app))
    ; comment out when testing == pretty bad workflow
    (p/show @ui)))

(plugin/defplugin lab.core.ui
  "Creates the UI for the application and hooks into
basic file operations."
  :init!    #'init!
  :hooks    hooks
  :keymaps  keymaps)

(comment

(do
  (def ui
    (let [app (init {:name "Clojure Lab - UI dummy"})
          ui  (app :ui)]
      ui))
  (def stylesheet {:split {:border :none}
                   :tabs  {:border :none}
                   :text-editor {:background 0x555555
                                 :font       [:name "Monospaced.plain" :size 14]}})
  (css/apply-stylesheet x stylesheet)
  (p/show @ui)
  nil)

)